atom/browser/api/atom_api_desktop_capturer.cc

```diff
@@ -82,7 +94,50 @@ void DesktopCapturer::OnSourceThumbnailChanged(int index) {
 }
 
 bool DesktopCapturer::OnRefreshFinished() {
-  Emit("finished", media_list_->GetSources());
+  const auto media_list_sources = media_list_->GetSources();
+  std::vector<DesktopCapturer::Source> sources;
+  for (const auto& media_list_source : media_list_sources) {
+    sources.emplace_back(
+        DesktopCapturer::Source{media_list_source, std::string()});
+  }
+
+#if defined(OS_WIN)
+  // Gather the same unique screen IDs used by the electron.screen API in order
+  // to provide an association between it and desktopCapturer/getUserMedia.
+  // This is only required when using the DirectX capturer, otherwise the IDs
+  // across the APIs already match.
+  if (using_directx_capturer_) {
+    std::vector<std::string> device_names;
+    // Crucially, this list of device names will be in the same order as
+    // |media_list_sources|.
+    webrtc::DxgiDuplicatorController::Instance()->GetDeviceNames(&device_names);
+    int device_name_index = 0;
+    for (auto& source : sources) {
+      if (source.media_list_source.id.type ==
+          content::DesktopMediaID::TYPE_SCREEN) {
+        const auto& device_name = device_names[device_name_index++];
+        std::wstring wide_device_name;
+        base::UTF8ToWide(device_name.c_str(), device_name.size(),
+                         &wide_device_name);
+        const int64_t device_id =
+            display::win::DisplayInfo::DeviceIdFromDeviceName(
+                wide_device_name.c_str());
+        source.screen_api_id = base::Int64ToString(device_id);
+      }
+    }
+  }
+#elif defined(OS_MACOSX)
+  // On Mac, the IDs across the APIs match.
+  for (auto& source : sources) {
+    if (source.media_list_source.id.type ==
+        content::DesktopMediaID::TYPE_SCREEN) {
+      source.screen_api_id =
+          base::Int64ToString(source.media_list_source.id.id);
+    }
+  }
+#endif  // defined(OS_WIN)
```

Comment:
It doesn't seem right to have empty screen IDs on Linux. It looks like the code here for macOS might also work on Linux (at least, it's not immediately obvious that there's anything mac-specific there); would it make sense to change the `#elif` to `#else`?

Reply:
The way I left it was indeed confusing; added a clarifying comment. The IDs don't match on Linux, but since Chrome can only capture the entire desktop currently, there's not much value in figuring it out now.